Documentation/writing-subplugin-tensor-filter.md

```diff
@@ -8,9 +8,60 @@ You can support a new neural-network framework (e.g., Tensorflow and Caffe) or a
 
 It is called "**subplugin**" because it is a plugin for a GStreamer plugin, ```tensor_filter```.
 
+**A subplugin should NEVER require properties of input/output dimensions or types.** You should use such properties for pipeline debugging and validation and the subplugin should fetch such information directly from model files via framework APIs and GSTCAPS negotiations.
 
 
-## Quick guide on writing a tensor\_filter subplugin for a new framework/hardware.
+
+# Quick guide on writing a tensor\_filter subplugin for a new framework/hardware.
+
+The most recommended method is to write a C++ tensor\_filter subplugin as a derived class of ```tensor_filter_subplugin``` class [/gst/nnstreamer/include/nnstreamer_cppplugin_api_filter.hh].
+Then, you can register the derived class, usually by calling ```tensor_filter_subplugin::register_subplugin()``` in the init function so that subplugin infrastructure recognizes your subplugin whtn the shared library is loaded.
+
+
+## A tensor\_filter subplugin C++ class
+
+Interface: [/gst/nnstreamer/include/nnstreamer_cppplugin_api_filter.hh]
+Reference (example): [/ext/nnstreamer/tensor_filter/tensor_filter_snap.cc]
+
+You may also find other subplugins inheriting the ```tensor_filter_subplugin``` class in [/ext/nnstreamer/tensor_filter] that can be used as examples.
+
+
+If the framework or backend/runtime library has C++ APIs or you want to write the subplugin in C++, use ```#include <nnstreamer_cppplugin_api_filter.hh>``` and inherit the base class, ```nnstreamer::tensor_filter_subplugin```.
+With this interface, subplugin writers are supposed to write their own concrete class based on ```nnstreamer::tensor_filter_subplugin``` by filling up virtual methods (methods marked ```These should be filled/implemented by subplugin authores```).
+Mandatory methods to be filled are declared as pure virtual function and optional methods are declared as regular virtual function (```eventHandler```).
+
+
+As in C subplugin, the derived (concrete) class should be registered at init and unregistered at exit.
+Subplugin writers are supposed to use the static methods of the base class, ```register_subplugin()``` and ```unregister_subplugin()```; refer to the function ```init_filter_snap()``` and ```fini_filter_snap()``` in the reference example.
```

Comment:
```suggestion
Subplugin writers are supposed to use the static methods of the base class, ```register_subplugin()``` and ```unregister_subplugin()``` refer to the function ```init_filter_snap()``` and ```fini_filter_snap()``` in the reference example.
```

Reply:
This semicolon is intentional. It was to break the sentence into two parts.